chainio/interface.go

```diff
@@ -0,0 +1,50 @@
+package chainio
+
+// Blockbeat defines an interface that can be used by subsystems to retrieve
+// block data. It is sent by the BlockbeatDispatcher whenever a new block is
+// received. Once the subsystem finishes processing the block, it must signal
+// it by calling NotifyBlockProcessed.
+//
+// The blockchain is a state machine - whenever there's a state change, it's
+// manifested in a block. The blockbeat is a way to notify subsystems of this
+// state change, and to provide them with the data they need to process it. In
+// other words, subsystems must react to this state change and should consider
+// being driven by the blockbeat in their own state machines.
+type Blockbeat interface {
+	// NotifyBlockProcessed signals that the block has been processed. It
+	// takes an error resulted from processing the block, and a quit chan
+	// of the subsystem.
+	//
+	// NOTE: This method must be called by the subsystem after it has
+	// finished processing the block. Extreme caution must be taken when
+	// returning an error as it will shutdown lnd.
+	//
+	// TODO(yy): Define fatal and non-fatal errors.
+	NotifyBlockProcessed(err error, quitChan chan struct{})
+
+	// Height returns the current block height.
+	Height() int32
+
+	// DispatchConcurrent sends the blockbeat to the specified consumers
+	// concurrently.
+	DispatchConcurrent(consumers []Consumer) error
```

Comment:
> ah ok - looks like this is how a subsystem can create consumers from within itself and use this to pass the beat to those consumers?

Yes. Since sometimes there are subsystems living inside a subsystem, e.g., multiple `ChannelArbitrator`s inside a `ChainArbitrator`, so I created this helper methods to let subsystem easily dispatch blockbeats.

> is it correct then to say that Consumers must call NotifyBlockProcessed after any calls to Dispatch*?
> I assume only the first level consumer...

Correct. As long as it's a `Consumer`, it must call `NotifyBlockProcessed`, which is intentionally by design.